### PR TITLE
Fix Legacy Widget block in Customizer

### LINKED ISSUE #31094: There are no legacy widgets in the Customizer
Here is a video showing the problem:

https://user-images.githubusercontent.com/107534/115732767-f373c200-a390-11eb-9837-3dab6b1cbde2.mp4

Problems:

- [ ] the legacy widgets provided by a plugin are not visible in the Customizer as variations of the `LegacyWidget`
- [ ] when I insert an empty `LegacyWidget` the Customizer crashes

### PR DESCRIPTION
**Pull request by noisysocks**
## Description

Fixes https://github.com/WordPress/gutenberg/issues/31094.

- Prevent crash when Legacy Widget block is added to Customizer.
- Show Legacy Widget block variations in Customizer.
- Prevent hidden widget types from appearing in Legacy Widget block dropdown in Customizer.
- Moves common functionality in `edit-widgets` and `customise-widgets` to a new `widgets` package.

## How has this been tested?

1. Enable the Widgets screen in Customizer experiment.
2. Install and activate the SiteOrigin Widgets Bundle plugin.
2. Navigate to Appearance → Customize → Widgets.
3. Add a Legacy Widget block and select a widget. This should now work.
5. Open the inserter and select a SiteOrigin widget variation. This should also now work.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/customize-widgets/src/register-legacy-widget-variations.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+								instance: {},
+						  }
+						: {
+								id: widgetType.id,
+						  },
+				} ) )
+			);
+		}
+	} );
+}
```
Comment (noisysocks) [inline packages/customize-widgets/src/register-legacy-widget-variations.js]: This is exactly the same as what's in `@wordpress/edit-widgets` which makes me think it's time for `@wordpress/widgets`. Should the "Move to sidebar" button go there too @talldan? Or anything else?

Comment (talldan) [reply packages/customize-widgets/src/register-legacy-widget-variations.js]: Yeah it could do. I can't think of anything else.

Comment (noisysocks) [reply packages/customize-widgets/src/register-legacy-widget-variations.js]: OK done!

**Review by talldan — COMMENTED** [review]
Nice work. Tested and this fixes the bugs mentioned in the issue.

There's one thing I spotted. It might be unrelated to this change, but I can't reproduce in `trunk` (mostly because I can't insert a legacy widget there).

1. Install the SiteOrigin bundle plugin
2. Open an empty widget area in the customizer
3. Add a legacy widget block
4. Select SiteOrigin Button option from the dropdown
5. Quicky click into the next default block appender (not the + button but the area where the paragraph will be created).
6. Editor crashes with the following error

```
react-dom.82e849f1.js:11237 Uncaught TypeError: Cannot read property 'id' of null
    at sidebar-adapter.js:252
    at Array.map (<anonymous>)
    at SidebarAdapter._updateWidgets (sidebar-adapter.js:251)
    at SidebarAdapter.setWidgets (sidebar-adapter.js:275)
    at use-sidebar-block-editor.js:186
    at basicStateReducer (react-dom.82e849f1.js:15148)
    at updateReducer (react-dom.82e849f1.js:15270)
    at updateState (react-dom.82e849f1.js:15372)
    at Object.useState (react-dom.82e849f1.js:16084)
    at useState (react.af754d50.js:1583)
 ```
 
 This line of code:
 https://github.com/WordPress/gutenberg/blob/1390bd74eb747e19b1e9f6754c8292f28548bed8/packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js#L238
 
It seems like a race condition, as it only happens when clicking quickly.

**Comment by noisysocks** [comment]
Good catch! Fixed in 6bab707.

**Review by talldan — APPROVED** [review]
Works great, nice one!

**Pull request merged**